code/io/tree_hdf5.c

```diff
@@ -94,19 +94,50 @@ void load_tree_table_hdf5(int filenr)
   for(i = 1; i < Ntrees; i++)
     TreeFirstHalo[i] = TreeFirstHalo[i - 1] + TreeNHalos[i - 1];
 
- // H5Fclose(hdf5_file);
-
 }
 
+#define READ_TREE_PROPERTY(sage_name, hdf5_name, type_int, data_type) \
+{ \
+  snprintf(dataset_name, MAX_STRING_LEN - 1, "tree_%03d/%s", treenr, #hdf5_name);\
+  status = read_dataset(hdf5_file, dataset_name, type_int, buffer);\
+  if (status != EXIT_SUCCESS) \
+  {\
+    ABORT(0);\
+  }\
+  for (halo_idx = 0; halo_idx < NHalos_ThisTree; ++halo_idx)\
+  {\
+    Halo[halo_idx].sage_name = ((data_type*)buffer)[halo_idx];\
+  }\
+} \
+
+#define READ_TREE_PROPERTY_MULTIPLEDIM(sage_name, hdf5_name, type_int, data_type) \
+{ \
+  snprintf(dataset_name, MAX_STRING_LEN - 1, "tree_%03d/%s", treenr, #hdf5_name);\
+  status = read_dataset(hdf5_file, dataset_name, type_int, buffer_multipledim);\
+  if (status != EXIT_SUCCESS) \
+  {\
+    ABORT(0);\
+  }\
+  for (halo_idx = 0; halo_idx < NHalos_ThisTree; ++halo_idx)\
+  {\
+    for (dim = 0; dim < NDIM; ++dim)\
+    { \
+      Halo[halo_idx].sage_name[dim] = ((data_type*)buffer_multipledim)[halo_idx * NDIM + dim];\
+    } \
+  }\
+} \
+
+
 void load_tree_hdf5(int32_t filenr, int32_t treenr)
 {
 
   char dataset_name[MAX_STRING_LEN];
   int32_t NHalos_ThisTree, status, halo_idx, dim;
 
-  int32_t *buffer_int;
-  float *buffer_float;
-  long long *buffer_longlong;
+  double *buffer; // Buffer to hold the read HDF5 data.  
+                  // The largest data-type will be double. 
+
+  double *buffer_multipledim; // However also need a buffer three times as large to hold data such as position/velocity.
```

Comment:
Just keep a single buffer -- allocate 3 times the memory ( or `dim` times) for the largest variable type read in